lib/prime.rb

```diff
@@ -132,7 +209,8 @@ def method_added(method) # :nodoc:
   #   Upper bound of prime numbers. The iterator stops after it
   #   yields all prime numbers p <= +ubound+.
   #
-  def each(ubound = nil, generator = EratosthenesGenerator.new, &block)
+  #def each(ubound = nil, generator = EratosthenesGenerator.new, &block)
+  def each(ubound = nil, generator = NextPrimeGenerator.new, &block)
```

Comment:
The code here modifies the original `prime?` inplementation.
It first determine which is the smallest witnesses range for the number.
It then creates an array of all the witnesses for that range.
It then uses those witnesses to peform the `miller_rabin_test` on.

The method `prime1?` shows the form of the implementation.
It's performance is the same a the original `prime` implemenation.
However, no that the witnesses are found first we can do `prime2`.
Using the `parallel` gem, it can now perform the miller-rabin tests in parallel,
with a significant speedup for very large numbers, and no degradation in
performance for small numbers.  

Below are average typical times (in seconds) with Ruby 2.3.1.
laptop: System76 Gazelle, 3.5GHz I7-6700HQ cpu, 4 cores, 8 threads

```
2.3.1 > def tm; s = Time.now; yield; Time.now-s end
2.3.1 > n = 10**2500; tm{ (n + 11859).prime? }      # et al


           (10**2500 + 11859)      (10**2500 + 13671)

prime?             3.16                    3.19        
prime1?            3.16                    3.19
prime2?            0.95                    0.96
```

Below is the whole file.
```
require 'openssl'
require 'parallel'

class Integer
  # Returns true if +self+ passes Miller-Rabin Test on +b+
  def miller_rabin_test(b)             # b is a witness to test with
    return self >= 2 if self <= 3 
    return false unless 6.gcd(self % 6) == 1
    n = d = self - 1
    d >>= 1 while d.even?
    y = b.to_bn.mod_exp(d, self)       # y = (b**d) mod self
    while d != n && y != n && y != 1
      y = y.to_bn.mod_exp(2, self)     # y = (y**2) mod self
      d <<= 1
    end
    y == n || d.odd?
  end

  # Returns true if +self+ is a prime number, else returns false.
  def prime?
    return true if [2, 3, 5, 7].include? self
    return false unless self > 1 and 210.gcd(self % 210) == 1
    Prime::A014233.each do |pair|
      ix, mx = pair
      return false unless miller_rabin_test(ix)
      break if self < mx
    end
    self == 41 ? true : miller_rabin_test(41)
  end
  
  def prime1?
    return true if [2, 3, 5, 7].include? self
    return false unless self > 1 and 210.gcd(self % 210) == 1
    witnesses = [2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41]
    val = Prime::A014233.values.sort.detect {|v| v > self}
    witnesses.select! {|p| p <= Prime::A014233.key(val)} if val
    witnesses.each { |p| return false unless miller_rabin_test(p) }
    true
  end

  def prime2?
    return true if [2, 3, 5, 7].include? self
    return false unless self > 1 and 210.gcd(self % 210) == 1
    witnesses = [2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41]
    val = Prime::A014233.values.sort,detect {|v| v > self}
    witnesses.select! {|p| p <= Prime::A014233.key(val)} if val
    Parallel.each(witnesses) { |p| return false unless miller_rabin_test(p) }
    true
  end
end

class Prime
  # https://oeis.org/A014233
  #
  # Smallest odd number for which Miller-Rabin primality test
  # on bases <= n-th prime does not reveal compositeness.
  #
  A014233 = {
    2 => 2047,
    3 => 1373653,
    5 => 25326001,
    7 => 3215031751,
    11 => 2152302898747,
    13 => 3474749660383,
    17 => 341550071728321,
    19 => 341550071728321,
    23 => 3825123056546413051,
    29 => 3825123056546413051,
    31 => 3825123056546413051,
    37 => 318665857834031151167461
  }
end
```